### PR TITLE
Fix publisher advertiseImpl() and subscribeImpl() for compressed_image_transport, compressed_depth_transport and theora_image_transport

### PR DESCRIPTION
**Pull request by ivanpauno**
Required after https://github.com/ros-perception/image_common/pull/249 and https://github.com/ros-perception/image_common/pull/252.

None of the parameters of this plugins were being declared because of this error.

### CODE REVIEW

**Comment by ivanpauno** [comment]
cc @jacobperron

**Review by jacobperron — APPROVED** [review]
LGTM, thanks :bow:

**Comment by jacobperron** [comment]
CI is complaining about some of the `override` keywords you added. Can you double check if those are necessary? Maybe there's a difference between Humble and Rolling,

**Comment by ijnek** [comment]
Duplicate of #105 (Draft), but happy to close the other if this gets merged first.
These changes would comply with new changes in publisher_plugin and subscriber_plugin from https://github.com/ros-perception/image_common/pull/249 too :+1:

**Comment by ivanpauno** [comment]
> These changes would comply with new changes in publisher_plugin and subscriber_plugin from https://github.com/ros-perception/image_common/pull/249 too +1

Yes, this needs a new release of `image_common`.

Actually, that PR is the one that broke this, right?
I'm not sure if it's a good idea the deprecation warning added there, as it's breaking downstream behavior completely.

**Comment by ivanpauno** [comment]
> CI is complaining about some of the override keywords you added. Can you double check if those are necessary? Maybe there's a difference between Humble and Rolling,

This doesn't apply to humble, it's needed after https://github.com/ros-perception/image_common/pull/249.

**Comment by ijnek** [comment]
> Actually, that PR is the one that broke this, right?

That's the PR that would break this once the changes are released, yes. (Hence #105 still being a draft). Just to confirm, were you building both image_common and image_transport_plugins from source?

The original problem that https://github.com/ros-perception/image_common/pull/249 solved was the error message that showed up at run time:
```
[ERROR] [1659558188.442971154] [rqt_gui_cpp_node_1318445]: SubscriberPlugin::subscribeImpl with five arguments has not been overridden
```
showed up when you called advertiseImpl without options, so there was a PR that hacked around this (https://github.com/ros-perception/image_common/pull/243). In https://github.com/ros-perception/image_common/pull/243#issuecomment-1186926280, I suggested deprecating image transports that don't handle options correctly.

**Comment by ivanpauno** [comment]
> That's the PR that would break this once the changes are released, yes. (Hence https://github.com/ros-perception/image_transport_plugins/pull/105 still being a draft). Just to confirm, were you building both image_common and image_transport_plugins from source?

Yes, I'm building both from source.

**Comment by ivanpauno** [comment]
@jacobperron can you create a new branch for humble so this can be merged in the `ros2` branch?
Thanks!

**Comment by jacobperron** [comment]
`humble` branch created. Related rosdistro PR: https://github.com/ros/rosdistro/pull/34104

**Comment by eigendude** [comment]
I also needed this change to compile compressed depth image transport:

```
diff --git a/compressed_depth_image_transport/include/compressed_depth_image_transport/compressed_depth_publisher.h b/compressed_depth_image_transport/include/compressed_depth_image_transport/compressed_depth_publisher.h
index d67af74..b3feecc 100644
--- a/compressed_depth_image_transport/include/compressed_depth_image_transport/compressed_depth_publisher.h
+++ b/compressed_depth_image_transport/include/compressed_depth_image_transport/compressed_depth_publisher.h
@@ -53,7 +53,8 @@ protected:
   virtual void advertiseImpl(
           rclcpp::Node * node,
           const std::string &base_topic,
-          rmw_qos_profile_t custom_qos) override final;
+          rmw_qos_profile_t custom_qos,
+          rclcpp::PublisherOptions options) override final;
 
   virtual void publish(const sensor_msgs::msg::Image& message,
                        const PublishFn& publish_fn) const override final;
diff --git a/compressed_depth_image_transport/src/compressed_depth_publisher.cpp b/compressed_depth_image_transport/src/compressed_depth_publisher.cpp
index aca3741..4eb4678 100644
--- a/compressed_depth_image_transport/src/compressed_depth_publisher.cpp
+++ b/compressed_depth_image_transport/src/compressed_depth_publisher.cpp
@@ -60,10 +60,11 @@ namespace compressed_depth_image_transport
 void CompressedDepthPublisher::advertiseImpl(
   rclcpp::Node * node,
   const std::string& base_topic,
-  rmw_qos_profile_t custom_qos)
+  rmw_qos_profile_t custom_qos,
+  rclcpp::PublisherOptions options)
 {
   typedef image_transport::SimplePublisherPlugin<sensor_msgs::msg::CompressedImage> Base;
-  Base::advertiseImpl(node, base_topic, custom_qos);
+  Base::advertiseImpl(node, base_topic, custom_qos, options);
 
   node->get_parameter_or<int>("png_level", config_.png_level, kDefaultPngLevel);
```

**Comment by ivanpauno** [comment]
Thanks @eigendude, I didn't check `compressed_depth_image_transport`.
I will update the PR with that fixed.

**Comment by ivanpauno** [comment]
I needed some extra fixes, but this is working fine locally for me.
We need a new rolling release of `image_common` to get CI passing.

File: theora_image_transport/src/theora_publisher.cpp [hunk 1]
```diff
@@ -74,20 +74,13 @@ TheoraPublisher::~TheoraPublisher()
 void TheoraPublisher::advertiseImpl(
   rclcpp::Node *node,
   const std::string &base_topic,
-  uint32_t queue_size,
-  rmw_qos_profile_t custom_qos)
+  rmw_qos_profile_t custom_qos,
+  rclcpp::PublisherOptions options)
 {
   logger_ = node->get_logger();
-  // queue_size doesn't account for the 3 header packets, so we correct (with a little extra) here.
```
Comment (ijnek) [inline theora_image_transport/src/theora_publisher.cpp:81]: I see you've removed this, is this change needed? (I don't quite know what it's doing)

File: theora_image_transport/src/theora_subscriber.cpp [hunk 2]
```diff
@@ -69,20 +69,13 @@ void TheoraSubscriber::subscribeImpl(
   rclcpp::Node * node,
   const std::string &base_topic,
   const Callback & callback,
-  uint32_t queue_size,
-  rmw_qos_profile_t custom_qos)
+  rmw_qos_profile_t custom_qos,
+  rclcpp::SubscriptionOptions options)
 {
   logger_ = node->get_logger();
-  // queue_size doesn't account for the 3 header packets, so we correct (with a little extra) here.
```
Comment (ijnek) [inline theora_image_transport/src/theora_subscriber.cpp:76]: Same comment as the previous

File: theora_image_transport/src/theora_publisher.cpp (hunk 1 above)
Comment (ivanpauno) [reply theora_image_transport/src/theora_publisher.cpp:81]: The previous comment of why that was being done doesn't make sense at all.
Also `queue_size` is not more an argument of the function.

**Comment by clalancette** [comment]
@ros-pull-request-builder retest this please

**Review by clalancette — APPROVED** [review]

**Comment by mhubii** [comment]
still receiving this error

```
[ERROR] [1719225304.277762595] [image_republisher]: SubscriberPlugin::subscribeImpl with five arguments has not been overridden
```

for compressed depth plugin on Ubuntu 22.04 ROS 2 Humble
